custom/icds_reports/static/js/angular-services/locations.service.js

```diff
@@ -122,19 +131,191 @@ window.angular.module('icdsApp').factory('locationsService', ['$http', '$locatio
+                            });
+                            if (vm.preventShowingAllOption(sortedLocations)) {
+                                locationsCache[parentId] = sortedLocations;
+                            } else if (selectedLocation.user_have_access) {
+                                locationsCache[parentId] = [ALL_OPTION].concat(sortedLocations);
+                            } else {
+                                locationsCache[parentId] = sortedLocations;
+                            }
+                        }
+                    }
+
+                    var levelOfSelectedLocation = _.findIndex(vm.hierarchy, function(locationTypes) {
+                        return _.contains(locationTypes.map(function(x) { return x.name; }), selectedLocation.location_type_name);
+                    });
+                    vm.selectedLocations[levelOfSelectedLocation] = vm.selectedLocationId;
+                    vm.onSelect(selectedLocation, levelOfSelectedLocation);
+
+                    levelOfSelectedLocation -= 1;
+
+                    while(levelOfSelectedLocation >= 0) {
+                        var childSelectedId = vm.selectedLocations[levelOfSelectedLocation + 1];
+                        var childSelected = _.find(locations, function(location) {
+                            return location.location_id === childSelectedId;
+                        });
+                        vm.selectedLocations[levelOfSelectedLocation] = childSelected.parent_id;
+                        levelOfSelectedLocation -= 1;
+                    }
+
+                    var levels = [];
+                    window.angular.forEach(vm.levels, function (value) {
+                        if (value.id > selectedLocationIndex()) {
+                            levels.push(value);
+                        }
+                    });
+                    vm.groupByLevels = levels;
+                    vm.selectedLevel = selectedLocationIndex() + 1;
+                });
+            } else {
+                vm.locationPromise = this.getRootLocations().then(function(data) {
+                    locationsCache.root = [NATIONAL_OPTION].concat(data.locations);
+                });
+                vm.groupByLevels = vm.levels;
+            }
+            return locationsCache;
+        },
+
+        onSelectLocation : function(item, level, locationsCache, vm) {
+            this.resetLevelsBelow(level, vm);
+            if (level < 4) {
+                vm.locationPromise = this.getChildren(item.location_id).then(function (data) {
+                    if (item.user_have_access) {
+                        locationsCache[item.location_id] = [ALL_OPTION].concat(data.locations);
+                        vm.selectedLocations[level + 1] = ALL_OPTION.location_id;
+                    } else {
+                        locationsCache[$item.location_id] = data.locations;
+                        vm.selectedLocations[level + 1] = data.locations[0].location_id;
```

Comment:
Is it possible for `data.locations` to be empty?

Reply:
Only if a location has no children and since data only comes from the lowest level, not meaningfully. Like a state can have no districts yet uploaded, but the state would also have zero data so you would have no reason to drill down into it.